# Area de una esfera

## La formula para el area de una esfera es la siguiente:

      area = 4 * PI * radio ** 2

1. Crea una funcion que calcula el area de la esfera.
2. Usa el `Pi` de `math`

In [1]:
import math

def area_esfera(radio):
  """
  Calcula el area de una esfera dado su radio
  """
  area = 4 * math.pi * radio**2
  return area

radio = 3.5
area = area_esfera(radio)
print("El area de la esfera con radio", radio, "es", area)

El area de la esfera con radio 3.5 es 153.93804002589985


# Volumen de una esfera

      volumen = (4/3) * pi * radio ** 3

1. Funcion que recibe el radio y calcula el volumen.
2. Usa el `Pi`de `math`.

In [2]:
def volumen_esfera(radio1):
    """
    Calcula el volumen de una esfera
    """
    volumen = (4 / 3) * math.pi * radio1 ** 3 
    return volumen

radio1 = 2
volumen = volumen_esfera(radio1)
print("El volumen de la esfera con radio", radio1, "es", volumen)

El volumen de la esfera con radio 2 es 33.510321638291124


# Los gringos y sus unidades absurdas 

Cuando vivía en EEUU, una de las cosas que más problemas me dió fueron sus unidades desquiciadas: onzas, onzas fluidas, millas, yardas y un sinfín de absurdeces que hacen que una tarea cotidiana se convierta en un desafío hercúleo. 

La cosa, lamentablemente, no termina ahí. También usan subdivisiones de dichas unidades. 

Esas subdivisiones no pueden lógicamente, ser decimales. Normalmente usan octavos (tal cual). 

## Mi némesis particular: el octavo de milla 

Vas por la carretera y waze te pega un grito, diciendo que tienes que salir a la derecha dentro de `3 octavos de milla`. ¿Mande? Te pones a echar cuentas, y cuando terminas, ves que era justo ahí atrás. 

Para evitar esto, vas a hacer una función que transforma octavos de milla en metros. 

  1. Recibes el número de octavos de milla (`octavos_milla`).
  2. Tienes una variable dentro de tu fución con la correspondencia millas a metros (`1609.344 metros` por `milla`).
  3. Con eso calculas el total de metros y lo devuelves.
  4. Usa la función `round` de Python (está en el ámbito global) para devolver un número entero (a nadie le importan los cm).


In [3]:
import math

def octavos_milla(octavos): 
    """"
    Se convierte de fracciones a decimales
    """
    
    millas = octavos * 0.125  
    metros = millas * 1609.344  

    return round(metros)

octavos = 3
metros = octavos_milla(octavos)
print(octavos, "octavos de milla son", metros, "metros")

3 octavos de milla son 604 metros


# Edad 

1. Crea una función que recibe tu año de nacimiento y calcula tu edad en años.

        edad_anos = ano_actual - ano_nacimiento

2. Crea una función que hace lo mismo pero te devuelve la edad en días (olvídate de los años bisiestos)

        edad_dias = edad_anos * 365
  
  ¿Se podría reaprovechar al primera función a la hora de crear la segunda y no repetir código? `DRY`.

In [4]:
def ano_nacimiento(edad_en_anos):
    """
    En la función ano_nacimiento, se ha creado la variable ano_nacimiento 
    para almacenar el año de nacimiento y se ha agregado la declaración 
    return para que la función devuelva el valor calculado.
    """
    ano_en_curso = 2023
    ano_nacimiento = ano_en_curso - edad_en_anos
    return ano_nacimiento

def edad_dias(edad_en_anos):
    """
    En la función edad_dias, se ha corregido el nombre de la variable 
    ano_en_curso a edad_en_anos y se ha llamado a la función ano_nacimiento 
    para obtener el año de nacimiento. Luego se ha calculado la edad en 
    días y se ha devuelto el valor.
    """
    edad_anos = ano_nacimiento(edad_en_anos)
    edad_dias = edad_anos * 365
    return edad_dias

edad_en_anos = 47
dias_cumplidos = edad_dias(edad_en_anos)

print("Tengo", edad_en_anos, "años y", dias_cumplidos, "días")

Tengo 47 años y 721240 días


# BMI: índice de masa corporal 

La operación bikini se aproxima y lo primero es saber cúanto tienes que adelgazar. Para eso, nada mejor que el BMI. 

EL BMI es un índice (un número) que se obtiene a partir de tu peso y altura y según dicho valor, se sabe si están en tu peso ideal, por encima del mismo o por debajo (no te lo crees ni tú). 

La fórmula, usando Kg para el peso y m para la altura es: 

              bmi = peso / altura**2

1. crea la función bmi que recibe peso y altura (en Kg y m).

2. devuelve el bmi.

3. Reeferencia: https://www.cdc.gov/healthyweight/spanish/assessing/bmi/adult_bmi/metric_bmi_calculator/bmi_calculator.html

In [6]:
def bmi(peso, altura):    
    """
    Calcula el índice de masa corporal (BMI) dado el peso (en Kg) 
    y la altura (en m).
    """
    bmi = peso / altura ** 2
    return bmi

mi_bmi = round(bmi(103, 1.89))
print("Mi indice de masa corporal es", mi_bmi, "%")

Mi indice de masa corporal es 29 %


# Euros a Pesetas 

Una función para los nostálgicos: vamos a convertir precios en Euros a Pesetas. La tasa de conversión es: `166.386 Pesetas` por `Euro`. 

1. Crea la función `euro_to_pta` que recibe un número de euros y te devuelve la cantidad de Pesetas.

        peseta = euro * 166.386

2. Ahora crea la función `pta_to_euro` que hace lo opuesto. 

        euro = peseta / 166.386

3. ¿Podrías implementar `pta_to_euro` llamando a `euro_to_pta` (para reaprovechar código)?

4. Referencia calculadora de cambio:  https://www.cambioeuro.es/peseta/

In [7]:
def euro_to_pta (euro):
    """
    Convierte una cantidad en euros a pesetas.
    """
    peseta = euro * 166.386

    return peseta

def pta_to_euro (peseta):
    """
    Convierte una cantidad en pesetas a euros.
    """
    euro = peseta / 166.386

    return euro

mi_euro = 10
mi_peseta = euro_to_pta(mi_euro)
print(mi_peseta)

mi_otro_euro = pta_to_euro(mi_peseta)
print(mi_otro_euro)

1663.86
10.0


# Energía cinética 

Crea la función `cinetic_energy(mass, speed)` que acepta la masa y la velocidad de un objeto y devuelve la energía cinética del objeto, calculada con la fórmula: 

    E = (1/2)mv²**

 Referencia: https://www.calcuvio.com/energia-cinetica

In [8]:
def cinetic_energy(mass, speed):
    """
    La variable energy almacena el resultado del cálculo de la energía cinética, 
    y luego la función devuelve este valor.
    """
    energy = (1/2) * mass * speed ** 2 

    return energy

my_energy = cinetic_energy(5, 5)
print(my_energy)

62.5


---

# Funciones Reversibles 

Las funciones son mecanismos que transforman una cosa en otra: te llevan de `A` a `B`.

Por ejemplo, te llevan de (peso, altura) a bmi, o de radio a area, etc. 

Se dice que mapean su entrada a su salida A veces, se puede conseguir encontrar la función inversa de una cierta función: 

Si `f` es una función que te lleva de `A` a `B`, entonces `f_inv` es una función que te lleva de `B` a `A`, `f_inv` es la inversa de `f`. 

# Ejercicio 1. 

## La función ¿`euro_to_peseta` es reversible? 

  La función `euro_to_pta` y `pta_to_euro` son reversibles, lo que significa que si conviertes una cantidad de euros a pesetas y luego conviertes la cantidad resultante de pesetas a euros, deberías obtener la cantidad original de euros.
  
  Como se puede observar, se obtiene la cantidad en pesetas correspondiente a 10 euros (1663.86 pesetas) y se convierte nuevamente a euros, obteniendo la cantidad original de 10 euros.

# Ejercicio 2. 

## La función `abs` de Python ¿es reversible?
  
  La función `abs()` de Python no es reversible en el sentido de que no es posible recuperar el valor original a partir de su valor absoluto. 
  
  La función toma un número y devuelve su valor absoluto, es decir, la distancia de ese número al cero sin importar si es positivo o negativo. Por ejemplo, abs(-5) devuelve 5. 
  
  Sin embargo, no es posible saber si el número original era -5 o 5 solo a partir del resultado de la función `abs()`. Por lo tanto, no es posible invertir la función `abs()` para recuperar el número original.
  
# Ejercicio 3. 

## La función `bmi` es reversible?

  Sí, la función `bmi` es reversible ya que se puede obtener el peso a partir del BMI y la altura. Solo se necesita despejar la fórmula del BMI de la siguiente manera:

  `bmi = peso / altura ** 2`
  
  `peso = bmi * altura ** 2`

  Por lo tanto, conociendo el BMI y la altura, se puede calcular el peso original.

---
# Funciones reversibles estimables 

Son aquellas que, aunque no se puede obtnere de una función inversa de forma directa, si tenemos más información, podemos estimar un `input` probable a aprtir del `output`. 

Por ejermplo, la función `bmi`. 

  Si tenemos muchos datos de alturas y pesos promedios o cómo suelen ir emparejados, a partir de un bmi podríamos estimar posibles combinaciones de altura y peso que sean más probables.Esto es lo que hace el `Deep Learning y las redes neuronales`. 
  
# Funciones Irreversibles 

Son aquellas que son de un sólo sentido y no hay manera práctica de obtener su inversa. Tienen una enorme importancia en la Ciencia de la Computación, especialmente en la criptografía: 

*   Cada vez que ves en tu browser una `URL` que empieza por `https`, quiere decir que todo lo que te manda el servidor y lo que mandas tú está encriptado mediante una función irreversible. Por ejemplo, cuando pones datos de tu tarjeta al comprar algo. 

*   Otro caso, más sencillo de funciones irreversibles que veremso (por encima) son las funciones hash (más adelante).

# Ecuación de 2º Grado 

Tienes la siguiente ecuación de segundo grado (¿te acuerdas?): 

    y = x ** 2 + 3 * x - 2

1. Crea la función que calcula y recibe x. 

## ¿Reversible? 

### Las funciones (o ecuaciones) cuadráticas (2º grado) ¿son reversibles?

Las funciones cuadráticas no son, en general, reversibles, ya que para un mismo valor de la variable independiente (x), puede haber dos posibles valores de la variable dependiente (y).

Por ejemplo, para la función y = x^2, si x = 2, entonces y = 4, pero si x = -2, también se tiene que y = 4. En este caso, no se puede invertir la función para obtener un único valor de x a partir de un valor de y, ya que hay dos posibles valores de x para un mismo valor de y.

Sin embargo, si se restringe el dominio de la función, es decir, se considera solo un subconjunto de valores de x para los cuales la función es inyectiva, entonces sí sería posible definir una función inversa.

In [9]:
# 1. Crea la funcion que calcula y recibe x

def ecuacion_segundo_grado(x):
    y = x ** 2 + 3 * x - 2
    return y
resultado = ecuacion_segundo_grado(2) # resultado = ecuacion_segundo_grado(2)
print(resultado)


8


---
# Funciones Totales y Parciales 
 
 Tenemos estas dos funciones matemáticas y tienes que crear las funciones de Python equivalentes 
 
 1. Prueba ambas funciones con varios valores para asegurarte que funcionan bien. 

    a)  `f(x, y) = x ** 2 + y ** 2 / x − y`

        # Función a)
              def f_a(x, y):
                return x**2 + y**2 / x - y
    
    b)  `f(m, n) = m ** 2 − n ** 2 + n / m ∗ n`

        # Función b)
              def f_b(m, n):
                return m**2 - n**2 + n / (m * n)
 
 2. ¿Hay algún valor de `x` o `y` que haga que haga que `(a)` estalle?

    En la función `f_a`, si `x` e y son iguales, se produce una división por cero, lo que haría que la función `estalle`. Por lo tanto, se debe evitar que `x` e `y` sean iguales.
    
 3. ¿Hay algún valor de `m` o `n` que haga que `(b)` estalle?

    En la función `f_b`, si `m` o `n` son cero, se produce una división por cero, lo que haría que la función `estalle`. Por lo tanto, se debe evitar que `m` o `n` sean cero.
 
 4. En la función cuadrática anterior, ¿hay algún valor de `x` que haga que la función estalle?

    En la función cuadrática `y = x**2 + 3*x - 2`, no hay ningún valor de `x` que haga que la función `estalle`. Por lo tanto, se puede decir que es una función total.
 
    "Una función es total si acepta todos los posibles valores de sus parámetros." 
  
  La función cuadrática anterior es total: acepta números y traga con cualquier número posible. 
  
    "Una función es parcial si hay uno o más valores de sus parámetros que hacen que no pueda hacer su trabajo."
  
  ### Las funciones (a) y (b) son parciales.

    El término "parcial" se refiere a que estas funciones no están definidas para ciertos valores de sus parámetros.
  
  - Si `x` e `y` son iguales, a `f(x,y)` le dá un error.

    En el caso de la función `(a)`, si `x` e `y` son iguales, el denominador en el segundo término de la ecuación se vuelve cero, lo que hace que la función no esté definida en ese punto y resulte en un error.
  
  - Si `m` o `n` son cero, a `f(m,n)` le dá un error.

    En el caso de la función `(b)`, si `m` o `n` son cero, el denominador en el segundo término de la ecuación se vuelve cero, lo que también hace que la función no esté definida en ese punto y resulte en un error.

    Por lo tanto, estas funciones son parciales porque no están definidas para ciertos valores de sus parámetros y producirán errores si se intenta evaluarlas en esos puntos.

---

# Conclusión 

Lo ideal sería que todas nuestras funciones fuesen funciones totales, facilitaría mucho nuestra vida. 

Como bien sabrás, la vida no es fácil y habrá muchas funciones parciales de por medio. En esos casos, 

- Tendremos que identificar las condiciones excepcionales (los valores que causan el erro) 

- Cuando se den, tendremso que gestionar dicho error La gestión de errores se verá con calma más adelante, pero ya eres consciete de su necesidad.

# Ejercicio 

Repasa todas las funciones que has creado hasta ahora y clasifícalas como totales o parciales. Si son parciales, identifica las condiciones excepcionales. 

## Distancia entre dos puntos en el plano cartesiano.

Si tenemos dos puntos: 

    p1(x1,y1) 
    
    p2(x2,y2) 

Su distancia cartesiana sería la siguiente: 

    distancia = √((x2 - x1)² + (y2 - y1)²)

1. Crear una función que recibe la `x` e `y` del primer punto y la `x` e `y` del segundo punto.

2. Devuelve la distancia cartesiana.

In [10]:
# Crear una función que recibe la x e y del primer punto y la x e y del segundo punto.

def distancia_cartesiana(x1, y1, x2, y2):
    """
    Calcula la distancia cartesiana entre dos puntos dados por sus coordenadas
    (x, y)
    """
    distancia = ((x1 - x2) ** 2 + (y1 - y2) ** 2) ** 0.5
    return distancia

# Se definen los dos puntos

punto1 = (2, 3)
punto2 = (5, 7)

""" 
Se calcula la distancia cratesiana entre ellos, En el código anterior se 
utiliza la función distancia_cartesiana para calcular la distancia cartesiana 
entre dos puntos. La función toma como argumentos las coordenadas x e y de 
cada punto y devuelve la distancia calculada utilizando la fórmula mencionada.
"""

distancia = distancia_cartesiana(punto1[0], punto1[1], punto2[0], punto2[1])

"""
Se llama a la función distancia_cartesiana pasándole como argumentos las 
coordenadas x e y de cada punto, que se encuentran almacenadas en las variables 
punto1 y punto2. La distancia calculada será almacenada en la variable distancia.
"""

# Se imprime el resultado

print("La distancia entre los puntos", punto1, "y", punto2, "es", distancia)

La distancia entre los puntos (2, 3) y (5, 7) es 5.0


# Curiosidad 

Esta es la `distancia cartesiana`, pero hay otros tipos de distancias definidas entre puntos. 

La `distancia cartesiana` es la primera que te muestra Google Maps o Waze cunado pides que te lleve a algún sitio. Normalmente cuando pides que calcule la ruta, te llevas un chasco, porque te da otra distancia mayor.

## ¿¿Por qué?? 

> Porque la distancia cartesiana sólo vale si puedes volar. Si tienes que ir entre calles, tendrás que dar vueltas y eso incrementa la distancia.

>¿Por qué crees que Google hace eso?

> Es porque la distancia que calculan es la distancia de conducción o la distancia por carretera, no la distancia en línea recta entre los dos puntos.
La distancia de conducción toma en cuenta factores como las carreteras y las calles que debes seguir, la velocidad límite en diferentes tramos, el tráfico y otros obstáculos en el camino. Mientras tanto, la distancia cartesiana solo mide la distancia en línea recta entre dos puntos.
Por lo tanto, la distancia de conducción generalmente será mayor que la distancia cartesiana, ya que es la distancia real que tienes que recorrer para llegar a tu destino en lugar de una distancia en línea recta entre dos puntos.

Otra distancia, algo más realista para la navegación por ciudades, es la distancia manhattan. 

> La distancia Manhattan se calcula sumando las diferencias absolutas entre las coordenadas de los puntos en cada dimensión. En otras palabras, se suma la distancia horizontal y la distancia vertical entre los dos puntos. Matemáticamente, la distancia Manhattan entre dos puntos (x1, y1) y (x2, y2) es:

    d_manhattan = |x1 - x2| + |y1 - y2|


Averigua en la wikipedia qué es la distancia manhattan, por qué se llama así y cómo calcularla. `Crea la función Python correspondiente`.

> La distancia Manhattan, también conocida como la distancia del taxista, es una medida de la distancia entre dos puntos en un espacio bidimensional. Se llama así porque se asemeja a cómo un taxi tendría que moverse en una ciudad con bloques cuadrados.

In [11]:
# Este es el calculo de la distancia Manhattan:

def distancia_manhattan(x1, y1, x2, y2):
    """
    Calcula la distancia Manhattan entre dos puntos en un plano cartesiano.
    """
    return abs(x1 - x2) + abs(y1 - y2)

"""
Esta función toma cuatro parámetros: las coordenadas x e y del primer punto 
y las coordenadas x e y del segundo punto. Devuelve la distancia Manhattan 
entre los dos puntos.
"""

distancia = distancia_manhattan(2, 3, 5, 7)
print(distancia)

7


---
# Abstracción al rescate 

El concepto de abstracción era precisamente empaquetar una o varias cosas, olvidarnos de sus tripas (caja negra) y tratarla como una sóla entidad. 

Esa es la salida a nuestro problema: 


> Tenemos que buscar una forma de empaquetar las dos coordenadas en un solo concepto (punto) y tratarlo como una sola cosa.

Ya hemos visto un ejemplo de algo así en Python. Recuerda los módulos: 

1. Representan todas las `definiciones globales` que hay dentro de un fichero. 

2. Por ejemplo el módulo `math` aglutina todas esas definiciones y te permite tratarlas como una sola cosa, llamada `math`. 

3. Para acceder a alguno de sus componentes, usamos punto: `math.pi` 

**No estaría de más poder hacer algo así con los puntos** y acceder a ellos de la siguiente manera: `punto1.y` o `punto1.x`.
  
  La función `area_vert` tendría una pinta más sencilla:

        def area_vert(v1, v2, v3, v4):
          # Recibes los vertices que encapsulan o agrupan
          # las dos coordenadas "x" e "y"
          # Para hacer las cuentas, si necesitas accder a una
          # de ellas, harias "v1.x" o "v2.y", etc...

---

# Abstracciones.

Hemos visto dos abstraciones hasta ahora: 

  1. **Abstracción de Variable:** Pillo un dato, lo meto dentro de una cajita y le doy un nombre. Me permite reutilizar datos. 
  
  2. **Abstracción de Función:** Pillo una expresión, la meto en una cajita y le doy un nombre. Me permite reutilizar expresiones. 
  
  Ahora hemos visto una tercera:

>**Abstración de Objeto:** pillo varias variables (en ¡incluso varias funciones!), las meto en una cajita y le doy un nombre. Me permite reutilizar esos conjuntos de variables y funciones y acceda a ellas mediante un punto.
  
Si te fijas, una engloba a otra, como las muñecas rusas: 
  
En el fondo, **no son más que diferentes formas de empaquetar cosas** en cajas negras y olvidarnos un poco de su contenido y detalles internos. Es decir, vagancia productiva. 
  
  # Abstracción de Objeto.
  
  Más adelante ya veremos con detalle la abstracción de objeto con detalle. Sin embargo, ya has sentido en tus carnes la necesidad que nos ha llevado a inventarla. De momento seguiremos con la abstracción de variable y la abstracción de función. 
  
  # ¿Qué hemos aprendido?

  Las funciones pueden ser: 
  
  Reversibles 
  Irreversibles (buenas para cripto) 
  Reversibles estimables (carne de red neuronal) 
  Totales (chachi) 
  Parciales (hay que identificar los errores y gestionarlos, con calma más adelante) 

* Reversibles 
* Irreversibles (buenas para cripto) 
* Reversibles estimables (carne de red neuronal) 
* Totales (chachi) 
* Parciales (hay que identificar los errores y gestionarlos, con calma más adelante) 


  
Hay 3 tipos de abstraciones: 
  
* Variable 
* Función 
* Objeto (veremos con calma más adelante) 
* Los 3 casos son formas de vaguear, hacer más con menos y olvidarnos de detalles irrelevantes